onnxruntime/core/providers/dml/DmlExecutionProvider/src/DmlGraphFusionTransformer.cpp

```diff
@@ -0,0 +1,391 @@
+            graphViewer,
+            *m_providerImpl->GetInternalRegistrationInfoMap(), 
+            kernel_lookup,
+            m_providerImpl->GetSupportedDeviceDataTypeMask(),
+            graphNodePropertyMap, 
+            requiredInitializerMap);
+
+        // Create a map between each initialized tensor and the partition(s) it is part of.
+        auto initializerPartitionMap = GetInitializerToPartitionMap(graphViewer, partitions);
+
+        for (uint32_t partitionIndex = 0; partitionIndex < partitions.size(); ++partitionIndex)
+        {
+            auto& partition = partitions[partitionIndex];
+
+            if (partition->GetRootMergedPartition() != partition.get() ||
+                !partition->IsDmlPartition())
+            {
+                continue;
+            }
+
+            // Create a map which will store by name each initializer which should be transferred to the 
+            // partition.  This prevents OnnxRuntime from allocating GPU resources and uploading those initializers,
+            // so the partiton's kernel can do so.  In the process, it will pre-process weights while consuming a CPU
+            // backed resource, avoiding an extra set of GPU resources in memory.
+            // A shared pointer is used so the functor and contained initializer captures can be cheaply copied within ORT.
+            auto transferredInitializerMap = std::make_shared<std::unordered_map<std::string, onnx::TensorProto>>();
+            
+            if (partition->IsDmlGraphPartition())
+            {
+                // populate transferredInitializerMap
+                for (const auto& input : partition->GetInputs())
+                {
+                    const onnx::TensorProto* tensor = nullptr;
+                    if (graph.GetInitializedTensor(input, tensor))
+                    {
+                        // It's only safe to transfer tensors which are used by this partition alone.
+                        auto iter = initializerPartitionMap.find(tensor);
+                        assert(iter != initializerPartitionMap.end());
+                        if (iter->second.size() > 1)
+                        {
+                            if (requiredInitializerMap.find(input) != requiredInitializerMap.end())
+                            {
+                                // The kernel relies on this input to be initialized, and it should be small enough to copy
+                                // cheaply. FusedGraphKernel only handles constant CPU inputs through transferred initializers,
+                                // rather than ORT, to avoid mismatches in policy or implementation causing failures.
+                                (*transferredInitializerMap)[input] = *tensor;
+                            }
+
+                            continue;
+                        }
+                        (*transferredInitializerMap)[input] = *tensor;
```

Comment:
>(*transferredInitializerMap)[input] = *tensor; [](http://example.com/codeflow?start=24&length=46)

Use std::move here to avoid the extra copy of weights.  Please check peak memory usage after the change

Reply:
We can't use std::move here, because `graph.RemoveInitializedTensor()` checks the existence of the initializer and fails if it doesn't find it. 
I think the extra copy of weight will be there very momentarily because just after the line  `(*transferredInitializerMap)[input] = *tensor;`, `graph.RemoveInitializedTensor()` will erase it's own copy.

Please let me know if I am missing something here.

Reply:
This would still impact session creation and peak memory usage considerably.  Let's discuss a solution.